src/sequential/medium_encryptor.rs

```diff
@@ -18,31 +20,38 @@ pub const MAX: usize = 3 * MAX_CHUNK_SIZE;
 // An encryptor for data which will be split into exactly three chunks (i.e. size is between
 // `3 * MIN_CHUNK_SIZE` and `3 * MAX_CHUNK_SIZE` inclusive).  Only `close()` will actually cause
 // chunks to be stored.  Until then, data is held internally in `buffer`.
-pub struct MediumEncryptor<S: Storage + Send + Sync> {
+pub struct MediumEncryptor<S: Storage + Send + Sync + Clone> {
     pub storage: S,
     pub buffer: Vec<u8>,
     original_chunks: Option<Vec<ChunkDetails>>,
 }
 
 impl<S> MediumEncryptor<S>
 where
-    S: Storage + 'static + Send + Sync,
+    S: Storage + 'static + Send + Sync + Clone,
 {
     // Constructor for use with pre-existing `DataMap::Chunks` where there are exactly three chunks.
     // Retrieves the chunks from storage and decrypts them to its internal `buffer`.
     #[allow(clippy::new_ret_no_self)]
     pub async fn new(
-        mut storage: S,
+        storage: S,
         chunks: Vec<ChunkDetails>,
     ) -> Result<MediumEncryptor<S>, SelfEncryptionError> {
         debug_assert_eq!(chunks.len(), 3);
         debug_assert!(MIN <= chunks.iter().fold(0, |acc, chunk| acc + chunk.source_size));
         debug_assert!(chunks.iter().fold(0, |acc, chunk| acc + chunk.source_size) <= MAX);
 
         let mut data = Vec::with_capacity(chunks.len());
+        let mut get_futures = Vec::new();
         for (index, chunk) in chunks.iter().enumerate() {
             let pad_key_iv = utils::get_pad_key_and_iv(index, &chunks);
-            let chunk = storage.get(&chunk.hash).await?;
+            let mut storage = storage.clone();
+            get_futures
+                .push(async move { storage.get(&chunk.hash).await.map(|res| (res, pad_key_iv)) });
+        }
+        let results = join_all(get_futures.into_iter()).await;
+        for chunk in results {
+            let (chunk, pad_key_iv) = chunk?;
             let decrypted_chunk = utils::decrypt_chunk(&chunk, pad_key_iv)?;
```

Comment:
can this `decrypt` operation to be moved into the `futures` as well? i.e. the result is `decrypted_chunk` directly?
i.e. the async function does `get` and `decrypt` in one go